// @azure/communication-react/dist/dist-esm/react-composites/src/composites/CallComposite/styles/CallComposite.styles.js
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.
import { mergeStyles } from '@fluentui/react';
const DEFAULT_Z_INDEX = 1;
const mainScreenContainerStyle = {
    height: '100%',
    width: '100%'
};
/**
 * @private
 */
export const mainScreenContainerStyleDesktop = mergeStyles(Object.assign(Object.assign({}, mainScreenContainerStyle), { minWidth: '30rem', 
    // max of min-width of composite pages (Call page)
    minHeight: '22rem' // max height of min-height of composite pages (Configuration page)
 }));
/**
 * @private
 */
export const mainScreenContainerStyleMobile = mergeStyles(Object.assign(Object.assign({}, mainScreenContainerStyle), { minWidth: '17.5rem', 
    // max of min-width of composite pages (Call page)
    minHeight: '13rem' // max height of min-height of composite pages (Configuration page & Call page)
 }));
/**
 * @private
 * Drawer styles to be used to house the _DrawerComponent on top of other content on the screen.
 * @param zIndex: this defaults to DEFAULT_Z_INDEX if unset
 */
export const drawerContainerStyles = (zIndex = DEFAULT_Z_INDEX) => ({
    root: {
        position: 'absolute',
        top: 0,
        left: 0,
        width: '100%',
        height: '100%',
        // Any zIndex > 0 will work because this is the only absolutely
        // positioned element in the container.
        zIndex: zIndex
    }
});
